plugin/pkg/global/resourcereferencemanager/admission.go

```diff
@@ -319,6 +323,98 @@ func (r *ReferenceManager) Admit(ctx context.Context, a admission.Attributes, o
 				})
 			}
 		}
+	case core.Kind("CloudProfile"):
+		cloudProfile, ok := a.GetObject().(*core.CloudProfile)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert resource into CloudProfile object")
+		}
+		if utils.SkipVerification(operation, cloudProfile.ObjectMeta) {
+			return nil
+		}
+		if a.GetOperation() == admission.Update {
+			oldCloudProfile, ok := a.GetOldObject().(*core.CloudProfile)
+			if !ok {
+				return apierrors.NewBadRequest("could not convert old resource into CloudProfile object")
+			}
+
+			// getting Kubernetes versions that have been removed from the CloudProfile
+			removedKubernetesVersions := sets.StringKeySet(helper.GetRemovedVersions(oldCloudProfile.Spec.Kubernetes.Versions, cloudProfile.Spec.Kubernetes.Versions))
+
+			// getting Machine image versions that have been removed from the CloudProfile
+			removedMachineImageVersions := map[string]sets.String{}
+			for _, oldImage := range oldCloudProfile.Spec.MachineImages {
+				imageFound := false
+				for _, newImage := range cloudProfile.Spec.MachineImages {
+					if oldImage.Name == newImage.Name {
+						imageFound = true
+						removedMachineImageVersions[oldImage.Name] = sets.StringKeySet(helper.GetRemovedVersions(oldImage.Versions, newImage.Versions))
+					}
+				}
+				if !imageFound {
+					for _, version := range oldImage.Versions {
+						if removedMachineImageVersions[oldImage.Name] == nil {
+							removedMachineImageVersions[oldImage.Name] = sets.NewString()
+						}
+						removedMachineImageVersions[oldImage.Name] = removedMachineImageVersions[oldImage.Name].Insert(version.Version)
+					}
+				}
+			}
+
+			if len(removedKubernetesVersions) > 0 || len(removedMachineImageVersions) > 0 {
+				shootList, err1 := r.shootLister.List(labels.Everything())
+				if err1 != nil {
+					return apierrors.NewBadRequest(fmt.Sprintf("could not list shoots to verify Kubernetes and Machine image version: %s", err1.Error()))
+				}
+
+				var (
+					channel = make(chan error)
+					wg      sync.WaitGroup
+				)
+				wg.Add(len(shootList))
+
+				for _, s := range shootList {
+					shoot := s.DeepCopy()
+					go func(ctx context.Context, shoot *core.Shoot, cloudProfile *core.CloudProfile, removedKubernetesVersions sets.String, removedMachineImageVersions map[string]sets.String) {
+						defer wg.Done()
+
+						if shoot.Spec.CloudProfileName != cloudProfile.Name {
+							channel <- nil
+							return
+						}
+
+						if removedKubernetesVersions.Has(shoot.Spec.Kubernetes.Version) {
+							channel <- fmt.Errorf("unable to delete Kubernetes version %q from CloudProfile %q  - version is still in use by shoot %q/%q", shoot.Spec.Kubernetes.Version, shoot.Spec.CloudProfileName, shoot.Name, shoot.Namespace)
+						}
+						for _, worker := range shoot.Spec.Provider.Workers {
+							if worker.Machine.Image == nil {
+								channel <- nil
+								return
```

Comment:
What if following workers have an image set, which is removed? I think, you won't detect it, if you do an early exit here.
```suggestion
								continue
```

Reply:
thanks, good catch. I added a test case for that scenario.

Reply:
Hmm, sry. I don't see it. Also, if I replace `continue` with `return` no test fails.